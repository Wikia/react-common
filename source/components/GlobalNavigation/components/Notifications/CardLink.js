import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';

import NotificationsContext from '../../models/NotificationContext';

const CardLink = ({ track, model, children }) => {
    const { markAsRead, goToDestination } = useContext(NotificationsContext);
    const { type, isUnread, uri, latestEventUri } = model;
    const onClick = useCallback((event) => {
        event.preventDefault();

        track({
            action: 'click',
            category: 'on-site-notifications',
            label: type,
            value: isUnread ? 1 : 0,
        });

        if (isUnread) {
            markAsRead(uri, true)
                .then(() => {
                    goToDestination(latestEventUri);
                });
        } else {
            goToDestination(latestEventUri);
        }
    }, [markAsRead, track, model, goToDestination]);

    return (
        <a
            className="wds-notification-card__outer-body"
            href={latestEventUri}
            onClick={onClick}
        >
            {children}
        </a>
    );
};

CardLink.propTypes = {
    children: PropTypes.node.isRequired,
    model: PropTypes.shape().isRequired,
    track: PropTypes.func.isRequired,
};

export default CardLink;
