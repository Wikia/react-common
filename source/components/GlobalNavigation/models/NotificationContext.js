import React from 'react';

const NotificationsContext = React.createContext({});

export default NotificationsContext;

export const NotificationsProvider = NotificationsContext.Provider;
export const NotificationsConsumer = NotificationsContext.Consumer;
