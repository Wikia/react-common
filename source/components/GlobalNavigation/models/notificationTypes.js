export const notificationTypes = {
    discussionUpvotePost: 'discussion-upvote-post',
    discussionUpvoteReply: 'discussion-upvote-reply',
    discussionReply: 'discussion-reply',
    announcement: 'announcement',
};

export const isDiscussionReply = type => type === notificationTypes.discussionReply;
export const isDiscussionReplyUpvote = type => type === notificationTypes.discussionUpvoteReply;
export const isDiscussionPostUpvote = type => type === notificationTypes.discussionUpvotePost;
export const isAnnouncement = type => type === notificationTypes.announcement;
