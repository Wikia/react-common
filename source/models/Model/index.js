import { Record, List } from 'immutable';

import isPlainObject from 'lodash/isPlainObject';

const BaseModel = (defaultValues, name) => class extends Record(defaultValues, name) {
    /**
     * Build  from data, returns undefined if data is not an object
     *
     * @param   {Object|Model} data
     * @returns {Model}
     */
    static build(data) {
        if (!(isPlainObject(data) || data instanceof this)) {
            return undefined;
        }

        const recordInstance = new this();
        return recordInstance.merge(data);
    }

    /**
     * Always returns default (empty)
     *
     * @returns {Model}
     */
    static empty() {
        return this.build({});
    }

    /**
     * Build a List<> collection
     *
     * @param   {Object|Object[]} data
     * @returns {List<Model>|undefined}
     */
    static buildList(data) {
        if (!data) {
            return undefined;
        }

        return List(data.map(item => this.build(item)));
    }
};

export default BaseModel;
