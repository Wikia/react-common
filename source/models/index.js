/**
 * Import test - please put all the components here
 * so Jest can verify that all the components are
 * importable.
 */

// eslint-disable-next-line
export { default as Model } from './Model';
