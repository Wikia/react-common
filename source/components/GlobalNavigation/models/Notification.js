import merge from 'lodash/merge';
import get from 'lodash/get';

import { convertToTimestamp } from '../utils/isoTime';

import { notificationTypes } from './notificationTypes';

const defaultProps = {
    title: null,
    snippet: null,
    timestamp: null,
    communityName: null,
    type: null,
    isUnread: false,
    totalUniqueActors: 1,
    latestActors: null,
    uri: null,
};

function mapData(notificationData) {
    return {
        title: get(notificationData, 'refersTo.title'),
        snippet: get(notificationData, 'refersTo.snippet'),
        uri: get(notificationData, 'refersTo.uri'),
        latestEventUri: get(notificationData, 'events.latestEvent.uri'),
        timestamp: convertToTimestamp(get(notificationData, 'events.latestEvent.when')),
        communityName: get(notificationData, 'community.name'),
        communityId: get(notificationData, 'community.id'),
        isUnread: notificationData.read === false,
        totalUniqueActors: get(notificationData, 'events.totalUniqueActors'),
        latestActors: createActors(get(notificationData, 'events.latestActors')),
        type: getNotificationType(notificationData),
    };
}

function createActors(actors) {
    return actors.map(actor => merge({}, actor, { profileUrl: getProfileUrl(actor.username) }));
}

function getProfileUrl(name) {
    return `${window.location.origin}/wiki/User:${name}`;
}

function getNotificationType(apiData) {
    if (apiData.type === 'upvote-notification') {
        if (apiData.refersTo.type === 'discussion-post') {
            return notificationTypes.discussionUpvoteReply;
        }

        return notificationTypes.discussionUpvotePost;
    }

    if (apiData.type === 'replies-notification') {
        return notificationTypes.discussionReply;
    }

    if (apiData.type === 'announcement-notification') {
        return notificationTypes.announcement;
    }

    return null;
}

class Notification {
    constructor(data) {
        Object.keys(data).forEach((key) => {
            this[key] = data[key];
        });
    }

    static build(notificationData) {
        return new this(merge({}, defaultProps, mapData(notificationData)));
    }
}

export default Notification;
