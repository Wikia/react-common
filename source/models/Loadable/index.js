// eslint-disable-next-line no-restricted-imports
import Model from '../Model';

export const LOADABLE_STATES = Object.freeze({
    UNINITIALIZED: 'UNINITIALIZED',
    IN_PROGRESS: 'IN_PROGRESS',
    COMPLETED: 'COMPLETED',
    ERROR: 'ERROR',
});
const LOADABLE_STATES_VALUES = Object.values(LOADABLE_STATES);

const LOADABLE_MODEL_KEYS = Object.freeze({
    state: '_loadable_state',
});

const loadableSchema = {
    [LOADABLE_MODEL_KEYS.state]: LOADABLE_STATES.UNINITIALIZED,
};

const LoadableModel = (defaultValues, name) => class extends Model({
    ...defaultValues,
    ...loadableSchema,
}, name) {
    _getLoadingState() {
        return this.get(LOADABLE_MODEL_KEYS.state);
    }

    _withLoadingState(state) {
        if (LOADABLE_STATES_VALUES.includes(state)) {
            return this.set(LOADABLE_MODEL_KEYS.state, state);
        }

        console.error(`LoadableModel._withLoadingState called with incorrect param: "${state}". Allowed states: ${LOADABLE_STATES_VALUES.join(', ')}.`);

        return this;
    }

    isLoadingUninitialized() {
        return this._getLoadingState() === LOADABLE_STATES.UNINITIALIZED;
    }

    isLoadingInProgress() {
        return this._getLoadingState() === LOADABLE_STATES.IN_PROGRESS;
    }

    isLoadingCompleted() {
        return this._getLoadingState() === LOADABLE_STATES.COMPLETED;
    }

    isLoadingError() {
        return this._getLoadingState() === LOADABLE_STATES.ERROR;
    }

    /**
     * is loading completed or error?
     */
    isLoadingFinished() {
        return this.isLoadingCompleted() || this.isLoadingError();
    }

    withLoadingInProgress() {
        return this._withLoadingState(LOADABLE_STATES.IN_PROGRESS);
    }

    withLoadingCompleted() {
        return this._withLoadingState(LOADABLE_STATES.COMPLETED);
    }

    withLoadingError() {
        return this._withLoadingState(LOADABLE_STATES.ERROR);
    }
};

export default LoadableModel;
