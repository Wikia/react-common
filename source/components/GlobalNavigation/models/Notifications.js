import get from 'lodash/get';

import { convertToIsoString } from '../utils/isoTime';
import Notification from './Notification';

function getNextPageLink(data) {
    return get(data, '_links.next') || null;
}

class Notifications {
    unreadCount = 0;
    nextPage = 0;
    notifications = [];

    constructor(apiInstance) {
        this.api = apiInstance;
    }

    static build(apiInstance) {
        return new this(apiInstance);
    }

    loadFirstPage() {
        return this.loadPage('/notifications?contentType=discussion-upvote&contentType=discussion-post&contentType=announcement-target');
    }

    loadPage(pageLink) {
        return this.api
            .fetchFromOnSiteNotifications(pageLink)
            .then(response => {
                this.notifications = this.mapNotifications(response.notifications);
                this.unreadCount = this.getUpdatedUnreadCount();

                return getNextPageLink(response);
            });
    }

    markAsRead(notification, willUnloadPage) {
        if (!notification.isUnread) {
            return Promise.resolve();
        }

        return notification.markAsRead(willUnloadPage)
            .then(() => {
                this.unreadCount = this.getUpdatedUnreadCount();
            });
    }

    markAllAsRead() {
        const since = convertToIsoString(this.notifications[0].timestamp);

        return this.api.fetchFromOnSiteNotifications(`/notifications/mark-all-as-read`, {
                body: JSON.stringify({ since }),
                headers: {
                    'Content-Type': 'application/json'
                },
                method: 'POST',
            }).then(() => {
                this.notifications = () => {}; // todo
                this.unreadCount = 0;
            });
    },

    mapNotifications(notifications) {
        return (notifications || []).map(notification => Notification.build(notification, this.api));
    }

    getUpdatedUnreadCount() {
        return this.notifications.filter(notification => notification.isUnread).length;
    }
}

