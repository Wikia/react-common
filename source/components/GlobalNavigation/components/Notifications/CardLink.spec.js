import React, { useContext, useCallback } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';

import { notificationTypes } from '../../models/notificationTypes';

import CardLink from './CardLink';

const defaultProps = {
    track: () => null,
    model: {
        communityId: '1619010',
        communityName: 'Xkxd Wiki',
        isUnread: false,
        latestActors: [{}],
        latestEventUri: 'http://xkxd.wikia.com/d/p/3100000000000001096/r/3086787452863005635',
        snippet: 'snippet',
        timestamp: 1550663179,
        title: 'MOAR LIKES',
        type: notificationTypes.discussionReply,
        totalUniqueActors: 1,
        uri: 'http://xkxd.wikia.com/d/p/3100000000000001096',
    },
    children: <div>IMMA CHILD</div>,
};

const mockDefaultContext = { markAsRead: () => Promise.resolve(), goToDestination: () => null };

function renderComponent(props) {
    const renderer = new ShallowRenderer();
    const computedProps = merge({}, defaultProps, props);

    renderer.render(<CardLink {...computedProps} />);

    return renderer.getRenderOutput();
}

// TODO: remove mocks when enzyme/react-test-renderer will fully support hooks
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(() => mockDefaultContext),
    useCallback: jest.fn(),
}));

/* eslint-disable-next-line jest/no-hooks */
beforeEach(() => {
    jest.resetModules();
});

test('CardLink renders correctly with default props', () => {
    expect(renderComponent()).toMatchSnapshot();
});

test('CardLink calls track and goToDestinationMock onClick when notification has inUnread set to false', () => {
    const markAsReadMock = jest.fn().mockResolvedValue();
    const preventDefaultMock = jest.fn();
    const trackMock = jest.fn();
    const goToDestinationMock = jest.fn();
    const latestEventUriMock = 'some-latest-mock';
    useContext.mockImplementation(
        () => ({ ...mockDefaultContext, markAsRead: markAsReadMock, goToDestination: goToDestinationMock }),
    );
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({ track: trackMock, model: { latestEventUri: latestEventUriMock } });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('a').props().onClick({ preventDefault: preventDefaultMock });

    expect(preventDefaultMock).toBeCalledWith();
    expect(trackMock).toBeCalledWith(expect.any(Object));
    expect(goToDestinationMock).toBeCalledWith(latestEventUriMock);
    expect(markAsReadMock).not.toBeCalled();
});

test('CardLink calls track, markAsRead and goToDestinationMock onClick when notification has inUnread set to true', () => {
    const markAsReadMock = jest.fn().mockResolvedValue();
    const preventDefaultMock = jest.fn();
    const trackMock = jest.fn();
    const goToDestinationMock = jest.fn();
    const uriMock = 'some-mock';
    const latestEventUriMock = 'some-latest-mock';
    useContext.mockImplementation(
        () => ({ ...mockDefaultContext, markAsRead: markAsReadMock, goToDestination: goToDestinationMock }),
    );
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({
        track: trackMock,
        model: { isUnread: true, uri: uriMock, latestEventUri: latestEventUriMock },
    });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('a').props().onClick({ preventDefault: preventDefaultMock });

    expect(preventDefaultMock).toBeCalledWith();
    expect(trackMock).toBeCalledWith(expect.any(Object));
    expect(markAsReadMock).toBeCalledWith(uriMock, true);
});
