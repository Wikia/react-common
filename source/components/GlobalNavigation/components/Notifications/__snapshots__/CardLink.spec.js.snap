// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardLink renders correctly with default props 1`] = `
<a
  className="wds-notification-card__outer-body"
  href="http://xkxd.wikia.com/d/p/3100000000000001096/r/3086787452863005635"
>
  <div>
    IMMA CHILD
  </div>
</a>
`;
